### PR TITLE
Audit Optimizations #1805 and #1933

### PR DESCRIPTION
**Pull request by prb112**
- fhir-audit: Ensure HOSTNAME is obtained correctly and consistently
#1805
- Auditing work performed when auditing is turned off for the
fhir-server #1933

Signed-off-by: Paul Bastide <pbastide@us.ibm.com>

### CODE REVIEW

File: docs/src/pages/guides/FHIRServerUsersGuide.md [hunk 1]
```diff
@@ -2196,6 +2196,8 @@ This section contains reference information about each of the configuration prop
 |`fhirServer/audit/serviceProperties/mapper`|string|The AuditEventLog mapper that determines the output format - valid types are 'cadf' and 'auditevent'. 'auditevent' refers to the FHIR Resource AuditEvent, and 'cadf' refers to the Cloud logging standard.|
 |`fhirServer/audit/serviceProperties/load`|string|The location that the configuration is loaded from 'environment' or 'config'.|
 |`fhirServer/audit/serviceProperties/kafka`|object|A set of name value pairs used as part of the 'config' for publishing to the kafka service. These should only be Kafka properties.|
+|`fhirServer/audit/hostname`|string|A string used to identify the Hostname, useful in containerized environments|
```
Comment (lmsurpre) [inline docs/src/pages/guides/FHIRServerUsersGuide.md:2199]: In our Cloud deploy, we use a different approach.  We have a configDropin that defines a 
`<variable name="defaultHostName">` and then we set this via environment variable.
Not sure if its better/worse, but just something to consider.

File: fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java [hunk 2]
```diff
@@ -59,411 +58,521 @@
...
+     * @param startTime
+     *            - The start time of the metadata request execution.
+     * @param endTime
+     *            - The end time of the metadata request execution.
+     * @param responseStatus
+     *            - The response status.
      */
     public static void logOperation(HttpServletRequest request, String operationName, String resourceTypeName, String logicalId,
-                                    String versionId, Date startTime, Date endTime, Response.Status responseStatus) {
+        String versionId, Date startTime, Date endTime, Response.Status responseStatus) {
```
Comment (lmsurpre) [inline fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java:522]:
```suggestion
            String versionId, Date startTime, Date endTime, Response.Status responseStatus) {
```

File: fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java [hunk 3]
```diff
@@ -59,411 +58,521 @@
...
+     *            - The start time of the request execution.
+     * @param endTime
+     *            - The end time of the request execution.
+     * @param responseStatus
+     *            - The response status.
      * @return AuditLogEntry - an initialized audit log entry.
      */
     private static AuditLogEntry populateAuditLogEntry(AuditLogEntry entry, HttpServletRequest request, Resource resource,
-                                                 Date startTime, Date endTime, Response.Status responseStatus) {
+        Date startTime, Date endTime, Response.Status responseStatus) {
```
Comment (lmsurpre) [inline fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java:575]:
```suggestion
            Date startTime, Date endTime, Response.Status responseStatus) {
```

File: fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java [hunk 4]
```diff
@@ -59,411 +58,521 @@
...
     /**
      * Builds an audit log entry for a 'create' REST service invocation.
-     * @param request - The HttpServletRequest representation of the REST request.
-     * @param resource - The Resource object being created.
-     * @param startTime - The start time of the create request execution.
-     * @param endTime - The end time of the create request execution.
-     * @param responseStatus - The response status.
+     *
+     * @param request
+     *            - The HttpServletRequest representation of the REST request.
```
Comment (lmsurpre) [inline fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java:67]: Minor style comment:  I don't think we use the `-` in other places where we add the param description on a new line.
```suggestion
     *            The HttpServletRequest representation of the REST request.
```
This applies to all the other places where you've reformatted the javadoc in this class as well.

File: fhir-audit/src/main/java/com/ibm/fhir/audit/configuration/handlers/ComponentIpHandler.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+import java.util.StringJoiner;
+import java.util.logging.Logger;
+
+import com.ibm.fhir.config.FHIRConfigHelper;
+import com.ibm.fhir.config.FHIRConfiguration;
+
+/**
+ * Calculates the Component Ip and subsequent retrieval of the IPs
+ */
+public class ComponentIpHandler {
```
Comment (lmsurpre) [inline fhir-audit/src/main/java/com/ibm/fhir/audit/configuration/handlers/ComponentIpHandler.java:24]: Please see FHIRException.getUniqueId() and determine if there is common functionality that we can use.  At least we should be using the same approach in the two places I think.

File: fhir-audit/src/main/java/com/ibm/fhir/audit/configuration/handlers/ComponentIpHandler.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    private static final String IP = determineIP();
+
+    public ComponentIpHandler() {
+        // No Operation
+    }
+
+    /*
+     * determines the IP
+     */
+    private static String determineIP() {
```
Comment (lmsurpre) [inline fhir-audit/src/main/java/com/ibm/fhir/audit/configuration/handlers/ComponentIpHandler.java:37]: Please see FHIRException.getUniqueId() and determine if there is common functionality that we can use. At least we should be using the same approach in the two places I think.

Comment (prb112) [reply fhir-audit/src/main/java/com/ibm/fhir/audit/configuration/handlers/ComponentIpHandler.java:37]: The problem with getUniqueId is it uses a lookup to get the LocalHost and can hang, specifically on lookup of the main hostname. This is a bit more resilient and faster. It also returns all IP addresses, where as the uniqueId only returns one.

File: fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java (hunk 4 above)
Comment (prb112) [reply fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java:67]: I think it's artifact of the class author.

File: docs/src/pages/guides/FHIRServerUsersGuide.md (hunk 1 above)
Comment (prb112) [reply docs/src/pages/guides/FHIRServerUsersGuide.md:2199]: I added code to pick that off if its set

Comment (prb112) [reply docs/src/pages/guides/FHIRServerUsersGuide.md:2199]: marking as resolved

File: fhir-audit/src/main/java/com/ibm/fhir/audit/configuration/handlers/ComponentIpHandler.java (hunk 6 above)
Comment (prb112) [reply fhir-audit/src/main/java/com/ibm/fhir/audit/configuration/handlers/ComponentIpHandler.java]: I've moved the handler to core

File: fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java (hunk 3 above)
Comment (prb112) [reply fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java:575]: fixed in the updated code

**Review by lmsurpre — APPROVED** [review]
LGTM

**Pull request merged**